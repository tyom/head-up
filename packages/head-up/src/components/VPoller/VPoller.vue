<template>
  <div class="VPoller">
    <div
      v-if="statusVisible"
      :style="{animationDuration: `${intervalMs}ms`}"
      class="status"
    />
    <slot :result="result"/>
  </div>
</template>

<script>
import parseDuration from 'parse-duration';

export default {
  props: {
    endpoint: {
      type: String,
      required: true,
    },
    interval: {
      type: String,
      required: true,
    },
  },
  data() {
    return {
      statusVisible: false,
      timer: undefined,
      result: undefined,
      intervalMs: parseDuration(this.interval),
    };
  },
  mounted() {
    this.startTimer();
    this.update();
  },
  beforeDestroy() {
    this.stopTimer();
  },
  methods: {
    startTimer() {
      this.timer = setInterval(this.update, this.intervalMs);
      this.statusVisible = true;
    },
    stopTimer() {
      clearInterval(this.timer);
      this.statusVisible = false;
    },
    async update() {
      this.stopTimer();
      const { data } = await this.$http.get(this.endpoint);
      this.result = data;
      this.startTimer();
    },
  },
};
</script>

<style scoped>
.VPoller {
  position: relative;
}

.status {
  animation: grow infinite linear;
  top: 0;
  left: 0;
  position: absolute;
  background-image: linear-gradient(90deg, #17619c00, #17619c);
  height: 1px;
  z-index: 1;

  &::after {
    content: '';
    position: absolute;
    border-radius: 50%;
    right: 0;
    top: 0;
    height: 1px;
    width: 1px;
    background-color: #fff;
    box-shadow: 0 0 0 1px #fff8;
  }
}

@keyframes grow {
  from {
    width: 0;
  }
  to {
    width: 100%;
  }
}
</style>
