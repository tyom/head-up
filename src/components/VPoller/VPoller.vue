<template>
  <div class="VPoller">
    <div
      v-if="statusVisible"
      class="status"
      :style="{ backgroundSize: `${intervalPercent}% 100%` }"
    >
      <div class="status-head" :style="statusHeadStyles" />
    </div>
    <slot :result="result" />
  </div>
</template>

<script>
import parseDuration from 'parse-duration';

export default {
  props: {
    endpoint: {
      type: String,
      required: true,
    },
    query: {
      type: Object,
      default: () => {},
    },
    requests: {
      type: Array,
      default: () => [],
    },
    interval: {
      type: String,
      required: true,
    },
  },
  data() {
    return {
      statusVisible: false,
      timer: undefined,
      result: undefined,
      intervalMs: parseDuration(this.interval),
    };
  },
  computed: {
    statusHeadStyles() {
      return {
        width: `${this.intervalPercent}%`,
        animationDuration: `${this.intervalMs}ms, ${this.intervalMs /
          this.intervalSeconds}ms`,
        animationTimingFunction: `steps(${this.intervalSeconds}, end), ease`,
      };
    },
    intervalSeconds() {
      return this.intervalMs / 1000;
    },
    intervalPercent() {
      return 100 / this.intervalSeconds;
    },
  },
  mounted() {
    this.update();
    this.startTimer();
  },
  beforeDestroy() {
    this.stopTimer();
  },
  methods: {
    startTimer() {
      this.timer = setInterval(this.update, this.intervalMs);
      this.statusVisible = true;
    },
    stopTimer() {
      clearInterval(this.timer);
      this.statusVisible = false;
    },
    getData(url, query) {
      return this.$http
        .get(url, {
          params: query,
        })
        .then(x => x.data);
    },
    populateRequests(requests) {
      return requests.map(request => {
        const url = this.endpoint.replace(/\/$/, '');
        const path = request.path || '';

        if (typeof request === 'string') {
          return this.getData(url + request, this.query);
        }
        return this.getData(url + path, {
          ...this.query,
          ...request.query,
        });
      });
    },
    async update() {
      const requests = this.populateRequests(this.requests);

      try {
        if (requests.length) {
          this.result = await Promise.all(requests);
        } else {
          this.result = await this.getData(this.endpoint, this.query);
        }
      } catch (err) {
        console.error(err.message);
      }
    },
  },
};
</script>

<style scoped>
.VPoller {
  position: relative;
}

.status {
  top: 0;
  left: 0;
  right: 0;
  position: absolute;
  background-image: linear-gradient(
    90deg,
    transparent 2%,
    rgba(#259fff, 0.4) 2%,
    rgba(#259fff, 0.4) 100%
  );
  height: 1px;
  z-index: 1;
}

.status-head {
  animation: tick infinite linear, fade infinite backwards;
  min-width: 1px;
  position: absolute;
  left: 0;
  top: 0;
  height: 100%;
  background-color: #fff;
}

._paused .status {
  visibility: hidden;
}

@keyframes tick {
  from {
    left: 0;
  }
  to {
    left: 100%;
  }
}

@keyframes fade {
  from {
    opacity: 0;
  }
  50% {
    opacity: 1;
  }
  to {
    opacity: 0;
  }
}
</style>
