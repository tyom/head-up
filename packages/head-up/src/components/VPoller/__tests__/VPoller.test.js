import { shallowMount, createLocalVue } from '@vue/test-utils';
import VueAxios from 'vue-axios';
import axios from 'axios';
import VPoller from '../VPoller';

jest.mock('axios');
jest.useFakeTimers();
jest.spyOn(global.console, 'error');

const localVue = createLocalVue();
localVue.use(VueAxios, axios);

axios.get.mockResolvedValue({
  data: {
    foo: 'bar',
  },
});

let wrapper;

beforeEach(() => {
  jest.clearAllMocks();
});

describe('single request', () => {
  beforeEach(() => {
    wrapper = shallowMount(VPoller, {
      localVue,
      propsData: {
        endpoint: 'http://localhost/',
        query: { key: 'abc' },
        interval: '10s',
      },
    });
  });

  test('default render', () => {
    expect(wrapper).toMatchSnapshot();
    wrapper.destroy();
  });

  test('calls endpoint after each interval', () => {
    expect(axios.get).toHaveBeenCalledTimes(1);
    expect(axios.get).toHaveBeenCalledWith('http://localhost/', {
      params: { key: 'abc' },
    });
    expect(setInterval).toHaveBeenCalledTimes(1);

    jest.runOnlyPendingTimers();

    expect(axios.get).toHaveBeenCalledTimes(2);
    expect(setInterval).toHaveBeenCalledTimes(1);

    jest.runOnlyPendingTimers();

    expect(axios.get).toHaveBeenCalledTimes(3);
    expect(setInterval).toHaveBeenCalledTimes(1);
  });

  test('removes interval before destroy', () => {
    expect(setInterval).toHaveBeenCalledTimes(1);
    expect(clearInterval).not.toHaveBeenCalled();

    wrapper.destroy();

    expect(clearInterval).toHaveBeenCalledTimes(1);
  });

  test('handles request error', async () => {
    axios.get.mockImplementationOnce(() => Promise.reject(new Error('nope')));

    await wrapper.vm.update();

    expect(console.error).toHaveBeenCalledWith('nope');
  });
});

describe('multiple requests', () => {
  test('calls multiple endpoint requests - shorthand', () => {
    wrapper = shallowMount(VPoller, {
      localVue,
      propsData: {
        endpoint: 'http://localhost/',
        query: { key: 'abc' },
        requests: ['/one', '/two'],
        interval: '10s',
      },
    });

    expect(axios.get.mock.calls).toEqual([
      ['http://localhost/one', { params: { key: 'abc' } }],
      ['http://localhost/two', { params: { key: 'abc' } }],
    ]);
  });

  test('calls multiple endpoint requests - with queries', () => {
    wrapper = shallowMount(VPoller, {
      localVue,
      propsData: {
        endpoint: 'http://localhost/',
        query: { key: 'abc' },
        requests: [
          {
            path: '/one',
            query: { a: 'A' },
          },
          {
            path: '/two',
          },
          '/three',
          {
            query: { b: 'B' },
          },
        ],
        interval: '10s',
      },
    });

    expect(axios.get.mock.calls).toEqual([
      ['http://localhost/one', { params: { key: 'abc', a: 'A' } }],
      ['http://localhost/two', { params: { key: 'abc' } }],
      ['http://localhost/three', { params: { key: 'abc' } }],
      ['http://localhost', { params: { key: 'abc', b: 'B' } }],
    ]);
  });
});
