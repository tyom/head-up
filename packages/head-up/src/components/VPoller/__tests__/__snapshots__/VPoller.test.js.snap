// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`single request default render 1`] = `
<div class="VPoller">
  <div class="status" style="background-size: 10% 100%;">
    <div class="status-head" style="width: 10%;"></div>
  </div>
</div>
`;
